File: Priority Queue/CourseSchedule3.cs
/*
 * Problem: https://leetcode.com/problems/course-schedule-iii/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int ScheduleCourse(int[][] courses) {
        Array.Sort<int[]>(courses, delegate(int[] a, int[] b) {
            if (a[1] == b[1]) return a[0] - b[0];
            return a[1] - b[1];
        });
        
        var current = 0;
        var PriorityQueue = new List<int>();
        foreach(var course in courses) {
            int duration = course[0];
            int end = course[1];
            
            // foreach course, put the duration in ordered list
            // then calculate the time, if the time exceed, pop the biggest duration
            var index = PriorityQueue.BinarySearch(duration);
            
            if (index < 0) index = ~index;
            PriorityQueue.Insert(index, duration);
            current += duration;
            
            if (current > end) {
                current -= PriorityQueue[PriorityQueue.Count - 1];
                PriorityQueue.RemoveAt(PriorityQueue.Count - 1);
            }
        }
        
        return PriorityQueue.Count;
    }
}

File: Javascript/CourseSchedule3.js
/**
 * Link: https://leetcode.com/problems/course-schedule-iii/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[][]} courses
 * @return {number}
 */
var scheduleCourse = function(courses) {
    // sort by the end date
    courses.sort(function(a, b) {
        if (a[1] == b[1]) return a[0] - b[0];
        return a[1] - b[1];
    })

    var order = [];
    var time = 0
    for (var i = 0; i < courses.length; i++) {
        var course = courses[i];
        var duration = course[0];
        var deadline = course[1];
        
        // for every course, put the duration into queue, and calculate the time. if the time exceed, pop the longest one.
        var index = binarySearch(order, duration);
        order.splice(index, 0, duration);
        time += duration;
        
        if (time > deadline) {
            var removeLongest = order.pop();
            time -= removeLongest;
        }
    }
    
    return order.length;
};


function binarySearch(arr, target) {
    var left = 0;
    var right = arr.length - 1;
    
    while(left < right) {
        var mid = Math.floor((left + right) / 2);
        
        if (arr[mid] < target) left = mid + 1;
        else right = mid;
    }
    
    if (arr[left] < target) return left + 1;
    return left;
}